File: app/src/main/java/starbright/com/projectegg/features/userAccount/UserAccountFragment.java
package starbright.com.projectegg.features.userAccount;

import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import com.afollestad.materialdialogs.DialogAction;
import com.afollestad.materialdialogs.MaterialDialog;

import java.util.Objects;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import starbright.com.projectegg.MyApp;
import starbright.com.projectegg.R;

public class UserAccountFragment extends Fragment implements UserAccountContract.View {

    private static final String USER_ACCOUNT_BUNDLE = "USER_ACCOUNT_BUNDLE";

    @BindView(R.id.tv_login_title)
    TextView tvLoginTitle;

    @BindView(R.id.tv_navigation)
    TextView tvNavigation;

    @BindView(R.id.et_email)
    EditText etEmail;

    @BindView(R.id.et_password)
    EditText etPassword;

    @BindView(R.id.btn_confirm)
    Button btnConfirm;

    @BindView(R.id.progress_bar)
    ProgressBar progressBar;

    @BindView(R.id.root_layout)
    ScrollView rootLayout;

    @Inject
    SharedPreferences sharedPreferences;

    private UserAccountContract.Presenter mPresenter;
    private FragmentListener mFragmentListener;

    public static UserAccountFragment newInstance() {
        return new UserAccountFragment();
    }

    @Override
    public void onAttach(Context context) {
        MyApp.getAppComponent().inject(this);
        super.onAttach(context);
        mFragmentListener = (FragmentListener) context;
    }

    @Override
    public void onCreate(@Nullable Bundle state) {
        super.onCreate(state);
        new UserAccountPresenter(this, state == null || state.getBoolean(USER_ACCOUNT_BUNDLE));
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_user_account, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view,
                              @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mPresenter.start();
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean(USER_ACCOUNT_BUNDLE, mPresenter.isLoginAuthentication());
    }

    @Override
    public void onDetach() {
        mFragmentListener = null;
        super.onDetach();
    }

    @Override
    public void onDestroyView() {
        mPresenter.onDestroy();
        super.onDestroyView();
    }

    @OnClick({
        R.id.tv_navigation,
        R.id.btn_confirm
    })
    void onButtonClicked(View view) {
        switch (view.getId()) {
            case R.id.tv_navigation:
                mPresenter.updateUserAuthRole();
                break;
            case R.id.btn_confirm:
                mPresenter.handleAuthenticationButtonClicked(
                        etEmail.getText().toString(),
                        etPassword.getText().toString()
                );
                break;
        }
    }

    @Override
    public void setPresenter(UserAccountContract.Presenter presenter) {
        mPresenter = presenter;
    }

    @Override
    public void updateView(boolean isLogin) {
        if (isLogin) {
            tvLoginTitle.setText(R.string.login_title);
            btnConfirm.setText(R.string.login_button_text);
            tvNavigation.setText(R.string.login_navigation);
        } else {
            tvLoginTitle.setText(R.string.register_title);
            btnConfirm.setText(R.string.register_button_text);
            tvNavigation.setText(R.string.register_navigation);
        }
    }

    @Override
    public void navigateToSearchRecipePage() {
        mFragmentListener.navigateToIngredientsActivity();
    }

    @Override
    public void disableView() {
        tvNavigation.setEnabled(false);
        etPassword.setEnabled(false);
        btnConfirm.setEnabled(false);
        etEmail.setEnabled(false);
        etEmail.setEnabled(false);
    }

    @Override
    public void enableView() {
        tvNavigation.setEnabled(true);
        btnConfirm.setEnabled(true);
        etPassword.setEnabled(true);
        etEmail.setEnabled(true);
    }

    @Override
    public void showLoginErrorToast(String errorMessage) {
        Snackbar.make(rootLayout, errorMessage, Snackbar.LENGTH_SHORT)
                .setAction(R.string.general_retry_label, new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        mPresenter.handleAuthenticationButtonClicked(etEmail.getText().toString(),
                                etPassword.getText().toString());
                    }
                })
                .setActionTextColor(getResources().getColor(R.color.lightRed))
                .show();
    }

    @Override
    public void showVerificationEmailSentDialog() {
        showDialogWithPositiveButtonOnly(
                R.string.register_verification_dialog_success_title,
                R.string.register_verification_dialog_success_message,
                R.string.register_verification_dialog_positive,
                new MaterialDialog.SingleButtonCallback() {
                    @Override
                    public void onClick(@NonNull MaterialDialog dialog,
                                        @NonNull DialogAction which) {
                        mPresenter.handleVerificationEmailDialogClicked();
                    }
                }
        );
    }

    @Override
    public void showProgressBar() {
        progressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideProgressBar() {
        progressBar.setVisibility(View.GONE);
    }

    @Override
    public void setupProgressBar() {
        if (MyApp.isPreLolipop()) {
            final Drawable drawableProgress = DrawableCompat.wrap(
                    progressBar.getIndeterminateDrawable());
            DrawableCompat.setTint(
                    drawableProgress,
                    ContextCompat.getColor(Objects.requireNonNull(getActivity()), R.color.red)
            );
            progressBar.setIndeterminateDrawable(DrawableCompat.unwrap(drawableProgress));
        } else {
            progressBar.getIndeterminateDrawable()
                    .setColorFilter(ContextCompat.getColor(Objects.requireNonNull(getActivity()),
                            R.color.red), PorterDuff.Mode.SRC_IN);
        }
    }

    @Override
    public void showEmailEmptyErrorToast() {
        showToast(R.string.general_error_email_empty);
    }

    @Override
    public void showEmailFormatWrongErrorToast() {
        showToast(R.string.general_error_email_format);
    }

    @Override
    public void showPasswordErrorToast() {
        showToast(R.string.general_error_password_empty);
    }

    private void showToast(@StringRes int messageRes) {
        Toast.makeText(getActivity(), getString(messageRes), Toast.LENGTH_SHORT).show();
    }

    private void showDialogWithPositiveButtonOnly(@StringRes int titleResId,
                                                  @StringRes int messageResId,
                                                  @StringRes int positiveTextResId,
                                                  MaterialDialog.SingleButtonCallback callback
    ) {
        new MaterialDialog.Builder(getActivity())
                .title(titleResId)
                .content(messageResId)
                .contentColor(ContextCompat.getColor(getActivity(), R.color.black))
                .positiveText(positiveTextResId)
                .onPositive(callback)
                .canceledOnTouchOutside(false)
                .build()
                .show();
    }

    interface FragmentListener {
        void navigateToIngredientsActivity();
    }
}

File: app/src/main/java/starbright/com/projectegg/features/userAccount/UserAccountFragment.kt
/**
 * Created by Andreas on 7/10/2019.
 */

package starbright.com.projectegg.features.userAccount

import androidx.fragment.app.Fragment

class UserAccountFragment : Fragment(), UserAccountContract.View {
    override fun updateView(isLogin: Boolean) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun navigateToSearchRecipePage() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun disableView() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun enableView() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun setupProgressBar() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun showProgressBar() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun hideProgressBar() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun showEmailEmptyErrorToast() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun showEmailFormatWrongErrorToast() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun showPasswordErrorToast() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun showLoginErrorToast(errorMessage: String?) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun showVerificationEmailSentDialog() {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

//    @Inject
//    lateinit var sharedPreferences: SharedPreferences
//
//    private lateinit var mPresenter: UserAccountContract.Presenter
//    private var mFragmentListener: FragmentListener? = null
//
//    override fun onAttach(context: Context?) {
////        MyApp.getAppComponent().inject(this)
//        super.onAttach(context)
//        mFragmentListener = context as FragmentListener?
//    }
//
//    override fun onCreate(state: Bundle?) {
//        super.onCreate(state)
//        UserAccountPresenter(this, true)
//    }
//
//    override fun onCreateView(inflater: LayoutInflater,
//                              container: ViewGroup?,
//                              savedInstanceState: Bundle?): View? {
//        val view = inflater.inflate(R.layout.fragment_user_account, container, false)
//        ButterKnife.bind(this, view)
//        return view
//    }
//
//    override fun onViewCreated(view: View,
//                               savedInstanceState: Bundle?) {
//        super.onViewCreated(view, savedInstanceState)
//        mPresenter.start()
//    }
//
//    override fun onDetach() {
//        mFragmentListener = null
//        super.onDetach()
//    }
//
//    override fun onDestroyView() {
//        mPresenter.onDestroy()
//        super.onDestroyView()
//    }
//
//    @OnClick(R.id.tv_navigation, R.id.btn_confirm)
//    internal fun onButtonClicked(view: View) {
//        when (view.id) {
//            R.id.tv_navigation -> mPresenter.updateUserAuthRole()
//            R.id.btn_confirm -> mPresenter.handleAuthenticationButtonClicked(
//                    et_email.text.toString(),
//                    et_password.text.toString()
//            )
//        }
//    }
//
//    override fun setPresenter(presenter: UserAccountContract.Presenter) {
//        mPresenter = presenter
//    }
//
//    override fun updateView(isLogin: Boolean) {
//        if (isLogin) {
//            tv_login_title.setText(R.string.login_title)
//            btn_confirm.setText(R.string.login_button_text)
//            tv_navigation.setText(R.string.login_navigation)
//        } else {
//            tv_login_title.setText(R.string.register_title)
//            btn_confirm.setText(R.string.register_button_text)
//            tv_navigation.setText(R.string.register_navigation)
//        }
//    }
//
//    override fun navigateToSearchRecipePage() {
//        mFragmentListener!!.navigateToIngredientsActivity()
//    }
//
//    override fun disableView() {
//        tv_navigation.isEnabled = false
//        et_password.isEnabled = false
//        btn_confirm.isEnabled = false
//        et_email.isEnabled = false
//    }
//
//    override fun enableView() {
//        tv_navigation.isEnabled = true
//        btn_confirm.isEnabled = true
//        et_password.isEnabled = true
//        et_email.isEnabled = true
//    }
//
//    override fun showLoginErrorToast(errorMessage: String?) {
//        Snackbar.make(root_layout, errorMessage!!, Snackbar.LENGTH_SHORT)
//                .setAction(R.string.general_retry_label) {
//                    mPresenter.handleAuthenticationButtonClicked(
//                            et_email.text.toString(),
//                            et_password.text.toString()
//                    )
//                }
//                .setActionTextColor(resources.getColor(R.color.lightRed))
//                .show()
//    }
//
//    override fun showVerificationEmailSentDialog() {
//        showDialogWithPositiveButtonOnly(
//                R.string.register_verification_dialog_success_title,
//                R.string.register_verification_dialog_success_message,
//                R.string.register_verification_dialog_positive,
//                MaterialDialog.SingleButtonCallback { dialog, which -> mPresenter!!.handleVerificationEmailDialogClicked() }
//        )
//    }
//
//    override fun showProgressBar() {
//        progress_bar.visibility = View.VISIBLE
//    }
//
//    override fun hideProgressBar() {
//        progress_bar.visibility = View.GONE
//    }
//
//    override fun setupProgressBar() {
//        if (MyApp.isPreLolipop) {
//            val drawableProgress = DrawableCompat.wrap(
//                    progress_bar.indeterminateDrawable)
//            DrawableCompat.setTint(
//                    drawableProgress,
//                    ContextCompat.getColor(Objects.requireNonNull<FragmentActivity>(activity), R.color.red)
//            )
//            progress_bar.indeterminateDrawable = DrawableCompat.unwrap(drawableProgress)
//        } else {
//            progress_bar.indeterminateDrawable.setColorFilter(
//                    ContextCompat.getColor(Objects.requireNonNull<FragmentActivity>(activity),
//                            R.color.red), PorterDuff.Mode.SRC_IN)
//        }
//    }
//
//    override fun showEmailEmptyErrorToast() {
//        showToast(R.string.general_error_email_empty)
//    }
//
//    override fun showEmailFormatWrongErrorToast() {
//        showToast(R.string.general_error_email_format)
//    }
//
//    override fun showPasswordErrorToast() {
//        showToast(R.string.general_error_password_empty)
//    }
//
//    private fun showToast(@StringRes messageRes: Int) {
//        Toast.makeText(activity, getString(messageRes), Toast.LENGTH_SHORT).show()
//    }
//
//    private fun showDialogWithPositiveButtonOnly(@StringRes titleResId: Int,
//                                                 @StringRes messageResId: Int,
//                                                 @StringRes positiveTextResId: Int,
//                                                 callback: MaterialDialog.SingleButtonCallback
//    ) {
//        MaterialDialog.Builder(activity!!)
//                .title(titleResId)
//                .content(messageResId)
//                .contentColor(ContextCompat.getColor(activity!!, R.color.black))
//                .positiveText(positiveTextResId)
//                .onPositive(callback)
//                .canceledOnTouchOutside(false)
//                .build()
//                .show()
//    }
//
//    override fun onBackPressed(): Boolean {
//        return mPresenter.handleOnBackPressedClicked()
//    }
//
//    internal interface FragmentListener {
//        fun navigateToIngredientsActivity()
//    }
//
//    companion object {
//
//        fun newInstance(): UserAccountFragment {
//            return UserAccountFragment()
//        }
//    }
}
